### ISSUE TITLE
Embed overlay pin into Cargo.nix

### ISSUE THREAD

**Issue by psionic-k** [cargo2nix/cargo2nix#157]
There's little reason to require the user to supply their own overlay pins while baking the Cargo.nix.  It adds a lot of process overhead to simple packaging and in 99% of cases the right answer is to use the same overlay version as at the time of creating the Cargo.nix

**Comment by psionic-k** [comment]
Just ran into an issue with an odd version of the nixpkgs-mozilla overlay giving me a rustc that had missed libz.so.1 in it's output.  Reinforces the need to pin during Cargo.nix generation and default to the same pins as the version of Cargo2nix.  The error initially looked like I needed a native dep and only after shelling in was it apparent that rustc was broken

Sample error message:
```
[nix-shell:~/debugging/yansi-term-0.1.2]$ /nix/store/qzrglv2ys3k4549ykx3ymyig9rykb241-rust-1.48.0-2020-11-16-7eac88abb/bin/rustc --version
/nix/store/qzrglv2ys3k4549ykx3ymyig9rykb241-rust-1.48.0-2020-11-16-7eac88abb/bin/rustc: error while loading shared libraries: libz.so.1: cannot open shared object file: No such file or directory
```

**Comment by psionic-k** [comment]
We can go several steps farther here in terms of burying the noise for most projects

* Default to placing the Cargo.nix into the /nix directory like niv.  For multiple-workspace shells, it's helpful
* Rename Cargo.nix to match the workspace name, becoming `myPackageWorkspace.nix`
* Embed default pins to avoid `packageFun` to `Cargo.nix` & `nixpkgs-mozilla` version mismatch pitfalls
* Construct the package set with all default args
* Provide the arguments to `mkShell` via function on the workspace so the user doesn't have to construct it manually

The result would be `(callPackage nix/static-web-serverWorkspace.nix { }).static-web-server.bin` to for instance, obtain the static web server program.

If workspaces can expose their shell dependencies in a set that can be merged, then we can build a shell by consuming several workspaces:
```nix
{
  sources ? import ./nix/sources.nix,
  system ? builtins.currentSystem,
  crossSystem ? null, #(import sources.nixpkgs {}).lib.systems.examples.musl64,
}:
let
  # get our pins in scope for callPackage
  inherit (sources) cargo2nix nixpkgs nixpkgs-mozilla;
in let
  staticServerWorkspace = callPackage nix/static-web-serverWorkspace.nix {};
  cobaltWorkspace = callPackage nix/cobaltWorkspace.nix {};
in let
  # here we need some function to merge a shell attribute from both workspaces
  # collisions are expected.  Might be necessary to support some filtering
  shellAttrs = magicMerge [static-web-serverWorkspace.shell {} cobaltWorkspace.shell {}];
in {
  shell = pkgs.mkShell  { shellAttrs };
  static-server = static-web-serverWorkspace.static-web-server.bin;
  cobalt = cobaltWorkspace
}
```

**Comment by psionic-k** [comment]
A key motivation that is emerging is to obtain nixpkgs instantiated with multiple `crossSystem` configuration and potentially platform overrides.

`system` is always the builder's platform, and so the flake-utils style creation of output sets is fine.

For a convoluted example, consider a repository that embeds a wasm output into a web server that is built into a docker container.  The build / host offsets in this case will form a list several members long.  Constructing the nixpkgs to obtain a `rustBuilder` for the given build / host combination is annoying.

Another issue that comes up is that target overrides for the Rust compiler can clash with the platform conditionals in the Cargo.nix.  The default should be consistency, with warnings to make the user aware that they are using the "I know what I'm doing" compiler-only override mechanisms.

There is an issue in constructing the shell from just one package set.  It should be possible to express a workspace shell from multiple sets, merging the transitive build dependencies of each workspace's crates so that cargo will work for anything that builds with Rust in that entire repository.

What does this have to do with implying the version of nixpkgs?  I believe we need a lib style API that does not require the user to touch overlay construction so much.  Most other nix "libraries" adopt this style.  We can use the inputs as the default for nixpkgs.  If the user overrides the cargo2nix.inputs.nixpkgs from within the flake, they can still change nixpkgs without having to go back into manual overlay construction.

Outputting several partial expressions for composition would make it easier to express all of these cases succinctly:
* target overrides
* cross setup
* package overrides
* special builder overrides for tests, clippy etc
* merging several workspaces of rust crates into a single set of cargo build-time deps

I also think relevant to #248 is going to be adopting this API in a way that there is one builder for obtaining finished binaries & libs but push all of the tests and other expressions into other builders.  The dedicated builder could be more stabilized, meaning that evaluated derivations will contain less line noise, helping caching.

It appears natural that all of these changes will happen in one set of commits.
